x/version/READ_ME.md

```diff
@@ -0,0 +1,103 @@
+# `x/version`
+
+## Abstract
+
+The version module will increment the app version in the header per a predefined
+mapping of app versions to heights. This allows for application logic to be
+routed in the same binary, which enables single binary syncs and upgrades.
+
+It accomplishes this by wrapping the `BaseApp`'s `EndBlock` method and checking
+if the current height is equal to the height at which the app version should be
+incremented. If it is, then the app version is incremented and the new version
+is set in the header.
+
+```go
+func (app *App) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBlock) {
+	res = app.BaseApp.EndBlock(req)
+	ctx := app.GetContextForDeliverTx([]byte{1})
+	return appversion.EndBlocker(ctx, app.VersionKeeper, res)
+}
+
+// EndBlocker will modify the app version if the current height is equal to
+// a predefined height at which the app version should be changed.
+func EndBlocker(ctx sdk.Context, keeper Keeper, resp abci.ResponseEndBlock) abci.ResponseEndBlock {
+	newAppVersion := keeper.GetVersion(ctx)
+	if ctx.BlockHeader().Version.App != newAppVersion {
+		resp.ConsensusParamUpdates.Version = &coretypes.VersionParams{
+			AppVersion: newAppVersion,
+		}
+		// set the version in the application to ensure that tendermint is
+		// passed the correct value upon rebooting
+		keeper.versionSetter.SetProtocolVersion(newAppVersion)
+	}
+	return resp
+}
+```
+
+## State
+
+The state in the version module is abnormal because it can be modified by the
+party building the binary without breaking the app hash. This is because the
+state is not stored in the iavl tree, but rather as a simple predefined golang
+map. Its important to note that even though this state is not in the state
+store, that is can still cause consensus breaking changes. This is because that
+state controls at which height the application will increment the app version.
+
+```go
+type Keeper struct {
+	chainAppVersions map[string]ChainVersionConfig
+}
+```
+
+## Events
+
+TODO: Add events
+
+### Usage
+
+The standard usage of the version module is to load the app version height
+mappings that are embedded into the binary, but it is also possible to load
+custom mappings by passing including a `map[string]ChainVersionConfig` in the
+app options using the `version.CustomVersionConfigKey` key.
+
+Mappings are defined by a `map[uint64]int64` where the key is the app version
+and the value is the height at which the app version should be incremented. For
+example:
+
+```go
+vm := map[uint64]int64{
+		1: 0,
+		2: 112093,
+		3: 300442,
+		4: 420420,
+}
+```
+
+The application will convert this mapping into a HeightVersionConfig:
```

Comment:
it wouldn't let me commit the suggested change cause its outdated, but was included in

https://github.com/celestiaorg/celestia-app/commit/e751399692a3d08774c84f894e453956a64b325f

used `ChainVersionConfig` instead of Height range, because that's what it gets converted to, but let me know if that doesn't make sense